File: Leetcode/BFS/#210. Course Schedule II.cpp
class Solution {
public:
    int cnt;
    vector<bool>visit;
    vector<int>res;
    
    void dfs(vector<vector<int>>&adj, int curr){
        if(visit[curr]) return;
        visit[curr]=true;
        
        for(auto child: adj[curr])
             dfs(adj,child);
        
        res[cnt--]=curr;
        return;
    }
    
    bool isCycle(int v, vector<vector<int>>&adj,vector<int>&degree){
        queue<int>Q;
        int cycle_cnt=v;
        
        for(int i=0; i<v; i++)
            if(degree[i]==0) Q.push(i);
        
        while(!Q.empty()){
            auto cur = Q.front(); Q.pop(); cycle_cnt--;
            for(auto child: adj[cur]){
                if(!--degree[child]) Q.push(child);
            }
        }
        
        return cycle_cnt;
    }
    
    vector<int> findOrder(int v, vector<vector<int>>& pre) {
        cnt=v-1;
        vector<vector<int>>adj(v);
        visit.resize(v,false);
        res.resize(v);
        vector<int>degree(v,0);
        
        for(auto edge: pre){
            adj[edge[1]].push_back(edge[0]);
            degree[edge[0]]++;
        }
        
        if(isCycle(v,adj,degree)) return {};
        
        for(int i=0; i<v; i++){
            if(visit[i]) continue;
            visit[i]=true;
            for(auto child: adj[i])
                dfs(adj,child);
            res[cnt--]=i;
        }
        
        return res;
    }
};

File: Graph/#210. Course Schedule II.py
'''
TC: O(V + E)
SC: O(V^2)
'''
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        res = []
        adj = [[] for _ in range(numCourses)]
        indegrees = [0 for _ in range(numCourses)]
        Q = []

        for pre in prerequisites:
            adj[pre[1]].append(pre[0])
            indegrees[pre[0]] += 1

        for i in range(numCourses):
            if indegrees[i] == 0:
                Q.append(i)

        while Q:
            curr = Q[0]
            Q.pop(0)
            res.append(curr)

            for ngbr in adj[curr]:
                indegrees[ngbr] -= 1
                if indegrees[ngbr] == 0:
                    Q.append(ngbr)

        return res if len(res) == numCourses else []

        